File: bridge/ble/protocol.js
/**
 * K-WATCH BLE protocol encoding/decoding.
 * Pure functions, no external dependencies.
 * Packet format: 20 bytes, zero-padded, little-endian multi-byte values.
 */

const PACKET_SIZE = 20;
const PAYLOAD_SIZE = 17;

const CMD_TIME_SYNC = 0x01;
const CMD_VIBRATE = 0x04;
const CMD_TRIGGER_LOST = 0x05;
const CMD_BATTERY = 0x0b;
const CMD_NOTIFICATION = 0x12;
const CMD_WEATHER = 0x22;
const CMD_KEEPALIVE = 0x3a;

const RESP_EVENT = 0x06;
const RESP_BATTERY = 0x0b;
const RESP_KEEPALIVE = 0x3a;

const EVENT_FIND_PHONE = 0x01;
const EVENT_TAKE_PHOTO = 0x02;

/**
 * Encode a notification as a multi-packet sequence.
 * @param {string} title
 * @param {string} body
 * @param {number} [typeId=1] - ANCS notification type (1=SMS)
 * @returns {Buffer[]} Array of 20-byte Buffers
 */
function encodeNotification(title, body, typeId = 1) {
  const titleBytes = utf8Truncate(title || '', PAYLOAD_SIZE);
  const bodyBytes = Buffer.from(body || '', 'utf8');

  const bodyChunks = [];
  if (bodyBytes.length === 0) {
    bodyChunks.push(Buffer.alloc(0));
  } else {
    for (let i = 0; i < bodyBytes.length; i += PAYLOAD_SIZE) {
      bodyChunks.push(bodyBytes.subarray(i, i + PAYLOAD_SIZE));
    }
  }

  const totalPackets = 2 + bodyChunks.length;
  const packets = [];

  // Packet 1: Header
  const pkt1 = Buffer.alloc(PACKET_SIZE);
  pkt1[0] = CMD_NOTIFICATION;
  pkt1[1] = totalPackets;
  pkt1[2] = 1;
  pkt1[3] = 0x00;
  pkt1[4] = typeId & 0xff;
  packets.push(pkt1);

  // Packet 2: Title
  const pkt2 = Buffer.alloc(PACKET_SIZE);
  pkt2[0] = CMD_NOTIFICATION;
  pkt2[1] = totalPackets;
  pkt2[2] = 2;
  titleBytes.copy(pkt2, 3);
  packets.push(pkt2);

  // Packets 3+: Body chunks
  for (let i = 0; i < bodyChunks.length; i++) {
    const pkt = Buffer.alloc(PACKET_SIZE);
    pkt[0] = CMD_NOTIFICATION;
    pkt[1] = totalPackets;
    pkt[2] = 3 + i;
    bodyChunks[i].copy(pkt, 3);
    packets.push(pkt);
  }

  return packets;
}

/**
 * Encode a time sync command (0x01).
 * @param {number} [tzOffsetHours] - Defaults to local timezone offset
 * @returns {Buffer}
 */
function encodeTimeSync(tzOffsetHours) {
  if (tzOffsetHours === undefined) {
    tzOffsetHours = -(new Date().getTimezoneOffset() / 60);
  }
  // The watch displays the timestamp directly without applying the offset,
  // so we send local time (UTC + offset) as the timestamp value.
  const localTimestamp = Math.floor(Date.now() / 1000) + (tzOffsetHours * 3600);
  const pkt = Buffer.alloc(PACKET_SIZE);
  pkt[0] = CMD_TIME_SYNC;
  pkt.writeUInt32LE(localTimestamp, 1);
  pkt[5] = tzOffsetHours & 0xff;
  return pkt;
}

/** @returns {Buffer} */
function encodeKeepaliveResponse() {
  const pkt = Buffer.alloc(PACKET_SIZE);
  pkt[0] = CMD_KEEPALIVE;
  return pkt;
}

/** @returns {Buffer} */
function encodeBatteryRequest() {
  const pkt = Buffer.alloc(PACKET_SIZE);
  pkt[0] = CMD_BATTERY;
  return pkt;
}

/**
 * Parse a response from the device.
 * @param {Buffer} data
 * @returns {{ type: string, [key: string]: any }}
 */
function parseResponse(data) {
  if (!data || data.length < 2) {
    return { type: 'unknown' };
  }

  const respId = data[0];

  if (respId === RESP_EVENT) {
    const eventCode = data[1];
    if (eventCode === EVENT_TAKE_PHOTO) return { type: 'event', eventCode, action: 'ok' };
    if (eventCode === EVENT_FIND_PHONE) return { type: 'event', eventCode, action: 'no' };
    return { type: 'event', eventCode, action: 'other' };
  }

  if (respId === RESP_BATTERY) {
    return { type: 'battery', level: data[1], charging: !!data[2] };
  }

  if (respId === RESP_KEEPALIVE) {
    return { type: 'keepalive' };
  }

  return { type: 'unknown' };
}

/**
 * Truncate a string to maxBytes of UTF-8, safe at character boundaries.
 * @param {string} text
 * @param {number} maxBytes
 * @returns {Buffer}
 */
function utf8Truncate(text, maxBytes) {
  const buf = Buffer.from(text, 'utf8');
  if (buf.length <= maxBytes) return buf;
  // Truncate and re-encode to avoid splitting multi-byte chars
  let truncated = buf.subarray(0, maxBytes);
  const str = truncated.toString('utf8').replace(/\uFFFD$/, '');
  return Buffer.from(str, 'utf8');
}

// OpenWeatherMap condition ID → watch weather type
const OWM_TO_WATCH = {
  200: 21, 201: 21, 202: 22, 210: 21, 211: 21, 212: 22, 221: 22, 230: 7, 231: 7, 232: 7,
  300: 4, 301: 4, 302: 4, 310: 4, 311: 4, 312: 5, 313: 5, 314: 6, 321: 4,
  500: 5, 501: 5, 502: 6, 503: 6, 504: 6, 511: 11, 520: 5, 521: 5, 522: 6, 531: 6,
  600: 8, 601: 9, 602: 10, 611: 11, 612: 11, 613: 11, 615: 11, 616: 11, 620: 8, 621: 9, 622: 10,
  701: 12, 711: 15, 721: 12, 731: 15, 741: 12, 751: 15, 761: 15, 762: 15, 771: 17, 781: 20,
  800: 1,
  801: 3, 802: 3, 803: 3, 804: 3,
};

/**
 * Encode a weather update for the watch. Command 0x22.
 * @param {object} w
 * @param {number} w.temp - Current temperature (°C)
 * @param {number} w.tempMin - Min temperature (°C)
 * @param {number} w.tempMax - Max temperature (°C)
 * @param {number} [w.conditionId=800] - OpenWeatherMap condition ID
 * @param {number} [w.uvIndex=0] - UV index
 * @param {number} [w.pm25=0] - PM2.5 (µg/m³)
 * @param {number} [w.aqi=0] - Air Quality Index
 * @param {number} [w.index=0] - Day index (0=today, 1-4=forecast)
 * @returns {Buffer}
 */
function encodeWeather(w) {
  const ts = Math.floor(Date.now() / 1000);
  const weatherType = OWM_TO_WATCH[w.conditionId] || 0;
  const curTemp = Math.round(w.temp || 0);
  const minTemp = Math.round(w.tempMin || 0);
  const maxTemp = Math.round(w.tempMax || 0);
  const uvIndex = Math.round(w.uvIndex || 0);
  const pm25 = Math.round(w.pm25 || 0);
  const aqi = Math.round(w.aqi || 0);

  const pkt = Buffer.alloc(PACKET_SIZE);
  pkt[0] = CMD_WEATHER;
  pkt[1] = (w.index || 0) & 0xff;
  pkt.writeUInt32LE(ts, 2);
  pkt.writeUInt16LE(weatherType, 6);      // daytime weather
  pkt.writeUInt16LE(weatherType, 8);      // nighttime weather
  pkt[10] = minTemp & 0xff;
  pkt[11] = maxTemp & 0xff;
  pkt[12] = 0;                            // air quality category (0-5)
  pkt.writeUInt16LE(pm25, 13);
  pkt[15] = uvIndex & 0xff;
  pkt.writeUInt16LE(aqi, 16);
  pkt[18] = curTemp & 0xff;
  return pkt;
}

/**
 * Encode a vibration command. Sends both 0x05 (trigger lost) and 0x04 (vibrate).
 * @param {number} [intensity=10] - Vibration intensity (0-10)
 * @returns {Buffer[]} Two 20-byte Buffers to send sequentially
 */
function encodeVibrate(intensity = 10) {
  const pkt1 = Buffer.alloc(PACKET_SIZE);
  pkt1[0] = CMD_TRIGGER_LOST;
  pkt1[1] = intensity & 0xff;
  const pkt2 = Buffer.alloc(PACKET_SIZE);
  pkt2[0] = CMD_VIBRATE;
  pkt2[1] = intensity & 0xff;
  return [pkt1, pkt2];
}

module.exports = {
  PACKET_SIZE,
  PAYLOAD_SIZE,
  CMD_NOTIFICATION,
  CMD_KEEPALIVE,
  encodeNotification,
  encodeTimeSync,
  encodeKeepaliveResponse,
  encodeBatteryRequest,
  encodeWeather,
  encodeVibrate,
  parseResponse,
};

File: custom_components/kwatch/protocol.py
"""K-WATCH BLE protocol encoding/decoding.

Pure functions with no Home Assistant or bleak dependency.
Packet format: 20 bytes, zero-padded, little-endian multi-byte values.
"""

from __future__ import annotations

import struct
import time

from .const import (
    CMD_KEEPALIVE,
    CMD_NOTIFICATION,
    CMD_TIME_SYNC,
    CMD_BATTERY,
    EVENT_FIND_PHONE,
    EVENT_TAKE_PHOTO,
    PACKET_PAYLOAD_SIZE,
    PACKET_SIZE,
    RESP_BATTERY,
    RESP_EVENT,
    RESP_KEEPALIVE,
)


def encode_notification(
    title: str, body: str, type_id: int = 1
) -> list[bytes]:
    """Encode a notification as a multi-packet sequence.

    Returns a list of 20-byte packets ready to write to the TX characteristic.
    Protocol: 0x46, totalPackets, seqId, then payload.
    """
    title_bytes = _utf8_truncate(title or "", PACKET_PAYLOAD_SIZE)
    body_bytes = (body or "").encode("utf-8")

    # Split body into 17-byte chunks (at least 1 chunk even if empty)
    body_chunks: list[bytes] = []
    if len(body_bytes) == 0:
        body_chunks.append(b"")
    else:
        for i in range(0, len(body_bytes), PACKET_PAYLOAD_SIZE):
            body_chunks.append(body_bytes[i : i + PACKET_PAYLOAD_SIZE])

    total_packets = 2 + len(body_chunks)  # header + title + body chunks
    packets: list[bytes] = []

    # Packet 1: Header
    pkt = bytearray(PACKET_SIZE)
    pkt[0] = CMD_NOTIFICATION
    pkt[1] = total_packets
    pkt[2] = 1  # sequence ID
    pkt[3] = 0x00
    pkt[4] = type_id & 0xFF
    packets.append(bytes(pkt))

    # Packet 2: Title
    pkt = bytearray(PACKET_SIZE)
    pkt[0] = CMD_NOTIFICATION
    pkt[1] = total_packets
    pkt[2] = 2
    pkt[3 : 3 + len(title_bytes)] = title_bytes
    packets.append(bytes(pkt))

    # Packets 3+: Body chunks
    for idx, chunk in enumerate(body_chunks):
        pkt = bytearray(PACKET_SIZE)
        pkt[0] = CMD_NOTIFICATION
        pkt[1] = total_packets
        pkt[2] = 3 + idx
        pkt[3 : 3 + len(chunk)] = chunk
        packets.append(bytes(pkt))

    return packets


def encode_time_sync(tz_offset_hours: int | None = None) -> bytes:
    """Encode a time sync command (0x01).

    Uses the current Unix timestamp. tz_offset_hours defaults to local timezone.
    """
    now = int(time.time())
    if tz_offset_hours is None:
        tz_offset_hours = -time.timezone // 3600
    pkt = bytearray(PACKET_SIZE)
    pkt[0] = CMD_TIME_SYNC
    pkt[1:5] = struct.pack("<I", now)
    pkt[5] = tz_offset_hours & 0xFF
    return bytes(pkt)


def encode_keepalive_response() -> bytes:
    """Encode a keepalive response (0x3A)."""
    pkt = bytearray(PACKET_SIZE)
    pkt[0] = CMD_KEEPALIVE
    return bytes(pkt)


def encode_battery_request() -> bytes:
    """Encode a battery level request (0x0B)."""
    pkt = bytearray(PACKET_SIZE)
    pkt[0] = CMD_BATTERY
    return bytes(pkt)


def parse_response(data: bytes | bytearray) -> dict:
    """Parse a 20-byte response from the device.

    Returns a dict with at minimum a "type" key.
    """
    if not data or len(data) < 2:
        return {"type": "unknown", "raw": bytes(data) if data else b""}

    resp_id = data[0]

    if resp_id == RESP_EVENT:
        event_code = data[1]
        if event_code == EVENT_TAKE_PHOTO:
            return {"type": "event", "event_code": event_code, "action": "ok"}
        if event_code == EVENT_FIND_PHONE:
            return {"type": "event", "event_code": event_code, "action": "no"}
        return {"type": "event", "event_code": event_code, "action": "other"}

    if resp_id == RESP_BATTERY:
        return {
            "type": "battery",
            "level": data[1],
            "charging": bool(data[2]),
        }

    if resp_id == RESP_KEEPALIVE:
        return {"type": "keepalive"}

    return {"type": "unknown", "raw": bytes(data)}


def _utf8_truncate(text: str, max_bytes: int) -> bytes:
    """Encode a string to UTF-8 and truncate at a safe byte boundary."""
    encoded = text.encode("utf-8")
    if len(encoded) <= max_bytes:
        return encoded
    # Truncate and re-decode to avoid splitting a multi-byte character
    truncated = encoded[:max_bytes]
    return truncated.decode("utf-8", errors="ignore").encode("utf-8")
